lib/facter/postgres_dirs_on_start.rb

```diff
@@ -0,0 +1,28 @@
+# Record postgres bin and data directories so data migrations can be performed
+# if postgres is upgraded
+require 'puppet'
+Facter.add(:postgres_most_current_bin_dir) do
+  confine :osfamily => %w{Debian Redhat}
+  os = Facter.value(:osfamily)
+  setcode do
+    if os == 'Debian'
+      cmd = 'find /usr/lib/postgresql -maxdepth 2 -type d -wholename "/usr/lib/postgresql/[0-9]*.[0-9]*/bin" | sort -V | tail -n 1'
+    else
+      cmd = 'find /usr -type f -name "psql" | sort -V | tail -n 2 | head -n 1 | xargs dirname'
```

Comment:
Since this code is run on each agent run, it would be great to have that code as tight as possible. Is it not possible to have a more specific test here?

Also, most of this could be implemented in ruby without having to spawn shells and subprocesses.